lib/Dialect/ModArith/IR/ModArithTypes.td

```diff
@@ -0,0 +1,48 @@
+#ifndef LIB_TYPES_MODARITH_IR_MODARITHTYPES_TD_
+#define LIB_TYPES_MODARITH_IR_MODARITHTYPES_TD_
+
+include "lib/Dialect/ModArith/IR/ModArithDialect.td"
+
+include "mlir/IR/DialectBase.td"
+include "mlir/IR/BuiltinTypeInterfaces.td"
+include "mlir/IR/AttrTypeBase.td"
+
+class ModArith_Type<string name, string typeMnemonic>
+    : TypeDef<ModArith_Dialect, name> {
+  let mnemonic = typeMnemonic;
+}
+
+def ModArith_ModArith : ModArith_Type<"ModArith", "mod_arith"> {
+  let summary = "Integer type with modular arithmetic";
+  let description = [{
+    `modulus` is the modulus the arithmetic working with.
+
+    `modulus` should be specified as, for example, `65537 : i32`.
+    It is required that the underlying integer type should be larger than
+    the modulus for a few bits. This requirement eases the handling of
+    modulus of similar bitwidth as the underlying type.
+
+    For example, when `modulus == 2 ** 32 - 1`, the underlying type
+    for the modulus should be `i64`.
+
+    The integer type should typically be `i32`, `i64` or higher.
+
+    Note that when the integer type is not specified, `i64` is implicitly
+    specified.
+
+    Examples:
+    ```
+    !Zp1 = !mod_arith.mod_arith<modulus = 7> // implicitly being i64
+    !Zp2 = !mod_arith.mod_arith<modulus = 65537 : i32>
+    !Zp3 = !mod_arith.mod_arith<modulus = 536903681 : i64>
+    ```
+  }];
+  let parameters = (ins
+    "::mlir::IntegerAttr":$modulus
+  );
+  let assemblyFormat = "`<` struct(params) `>`";
```

Comment:
If we don't think we will have additional parameters, we could even try

```suggestion
  let assemblyFormat = "`<` $modulus `>`";
```

And it would skip the `modulus = ` part, I think.